### ISSUE TITLE
Accessing bin_data in AbsoluteIMU(-A/C/G)

### ISSUE THREAD

**Issue by cavenel** [ddemidov/ev3dev-lang-python#4]
Hi, and thank you for the great job on this binding!
I am trying to use the mindsensors.com Gyro, MultiSensitivity Accelerometer and Compass (http://www.ev3dev.org/docs/sensors/mindsensors.com-gyro-multisensitivity-accelerometer-and-compass/#fnref:all) from Python.
When it is in mode "ALL", it is supposed to give values via "bin_data" with a "bin_data_format". I am trying to find the best way to access and convert this "bin_data" buffer in your bindings. Do you have any ideas?

**Comment by ddemidov** [comment]
Hi, and thanks!

It does not look like `bin_data` attribute is supported by the C++ bindings that the python module is based upon. In fact, it looks like the attribute is missing from the unified bindings specification. So currently there is no way to access it directly.

I may try to implement the attribute in C++ and then expose it in python. But I am not sure yet how exactly to do this (the user interface part). The problem is the data does not have a fixed format (it could be an array of signed/unsigned chars, ints, or even floats). May be it would be best to just return an array of bytes and let the user decide what to do with it. What do you think?

**Comment by WasabiFan** [comment]
> The problem is the data does not have a fixed format (it could be an array of signed/unsigned chars, ints, or even floats). May be it would be best to just return an array of bytes and let the user decide what to do with it.

This is why I didn't include this in our spec -- I had no ideas on how it should be implemented. If you decide on how you think this should be done, I'd say we should add it to the spec in the next release.

**Comment by cavenel** [comment]
I think the solution of returning an array of bytes is the best. As you mentioned, it can contain any type of data, so let the user decide what to do with it. But I think it is important to have a way of accessing it somehow.

**Comment by ddemidov** [comment]
I've uploaded an egg that is built from `bin_data` branches here: http://ddemidov.github.io/ev3dev-lang-python/_downloads/python_ev3dev-bin_data.egg. Can you check that it works for you?

Here is what I get for `infrared_sensor`:
```python
>>> from ev3dev import *
>>> ir = infrared_sensor()
>>> ir.bin_data_format
's8'
>>> ir.bin_data
<read-only buffer ptr 0x3eeb68, size 1 at 0xb6a36c40>
>>> ord(ir.bin_data[0])
18
>>> ir.value()
18
>>> 
```

**Comment by cavenel** [comment]
Perfect!
Here is my utilisation for the mindsensors AbsoluteIMU:
```python
    from ev3dev import *
    import struct

    gyro = sensor("in1:i2c17")
    print "sensor connected:", gyro.connected
    print gyro.num_values, "==", len(gyro.bin_data)
    print struct.unpack(len(gyro.bin_data)*'B', gyro.bin_data)
    print [gyro.value(i) for i in range(8)]
```
    cavenel@dhcppc2:~$ python mindsensors.py 
    sensor connected: True
    23 == 23
    (242, 77, 92, 150, 3, 97, 254, 191, 254, 217, 0, 188, 3, 115, 3, 65, 1, 6, 0, 15, 0, 1, 0)
    [242, 77, 92, 150, 3, 97, 254, 191]

It is a bit weird that num_values = 23 when only 8 values are possible and the other are in bin_data, but that's a problem with the driver directly and is written is the doc.
Should I close the issue or do you want to wait for other comments?

Thank's again!

**Comment by ddemidov** [comment]
Great! It looks even worse than I thought though: you are unpacking 3 chars followed by 10 shorts. There is no way this could be exposed in the API, so providing an array of bytes is the way to go. @WasabiFan, what do you think?

The 23 here is probably just the size of `bin_data` in bytes, since that is the type used at the front of your structure. I would have no way of reading the correct amount of data without this.

**Comment by WasabiFan** [comment]
> There is no way this could be exposed in the API, so providing an array of bytes is the way to go. @WasabiFan, what do you think?

I think that we should definitely expose the plain array but there may be _a bit_ more that we could do.

What if we added methods that, like the sensor class, would return the value with different types? For example, maybe we'd add a version of the function that would get a float at a specified byte index. Do you think that that would be more helpful than having the users do it themselves? I know that in Node, getting a float from a specified position in a byte array requires ~2-3 lines of code, so maybe we could simplify that into a single call. Something like:

```js
sensor.getBinDataFloat(4);
```

**Comment by ddemidov** [comment]
> sensor.getBinDataFloat(4); 

I don't like this particular idea, because accessing each field in a binary structure would require another file IO operation. This would be both inefficient and potentially error prone, since the readings for different fields may become de-synchronized (come from different time points).

But you are correct in that each binding may have slightly different implementation of this attribute. Even for C++ `get_bin_data` gets special treating. So I would say the attribute does not belong to [direct attribute mappings](https://github.com/ev3dev/ev3dev-lang/blob/f9bdd2072f925ba25420e5c81e90339aae0e0993/wrapper-specification.md#direct-attribute-mappings-3) section of specification, but should be listed in [special properties](https://github.com/ev3dev/ev3dev-lang/blob/f9bdd2072f925ba25420e5c81e90339aae0e0993/wrapper-specification.md#special-properties-3) and each implementation should decide what is the best way to deal with it.

For example, in C++ users could supply a custom struct and just ask the `sensor` class to fill the struct with binary data:
```cpp
struct  __attribute__((__packed__)) abs_imu_data {
  char tilt[3];
  short accel[3];
  short compass;
  short mag[3];
  short gyro[3];
} val;

// The get_bin_data is implemented as template:
gyro.get_bin_data(val);
```

In python, I like @cavenel's approach with `struct.unpack`, so it could be something like
```python
values = gyro.get_bin_data('<bbbhhhhhhhhhh')
```

I don't know what is the best approach for javascript or lua, but I am sure it would be something different.

By the way, @cavenel, in your example above it looks like values returned by `struct.unpack` and `gyro.value(i)` have nothing common between them. Is this to be expected?

**Comment by cavenel** [comment]
You are right for the 23, it is important to extract bin_data.

According to (ev3dev documentation on the AbsoluteIMU sensor)[http://www.ev3dev.org/docs/sensors/mindsensors.com-gyro-multisensitivity-accelerometer-and-compass/], values don't even exist when mode = ALL. They apparently only convert the first 8 bytes of bin_data to unsigned char. They would correspond if I used ` struct.unpack('BBBBBBBB', gyro.bin_data[:8])` but that doesn't make any sense.

Here is some code to check what corresponds to what on this sensor:
```Python
from ev3dev import *
import struct, time, math

class AbsoluteIMU(sensor):
    def __init__(self):
        super(AbsoluteIMU, self).__init__("in1:i2c17")

    def get_all_values(self):
        self.mode = "ALL"
        v = list(struct.unpack('<bbbhhhhhhhhhh', self.bin_data))
        return {'TILT_RAW':v[:3], 'ACCEL':v[3:6], 'COMPASS':v[6], 'MAG':v[7:10], 'GYRO_RAW':v[10:13],
                'TILT': [360.*math.asin(t/128.)/(2.*math.pi) % 180 for t in v[:3]],
                'GYRO': [t*0.1 for t in v[10:13]]}
    
AIMU = AbsoluteIMU()

for mode in ["TILT","ACCEL","COMPASS","MAG","GYRO"]:
    AIMU.mode = mode
    time.sleep(0.1) # We have to give the sensor some time to change mode.
    print "Mode", mode,":\t",
    print [AIMU.value(i) for i in range(AIMU.num_values)], AIMU.get_all_values()[mode]
    time.sleep(0.01)

print "Show 8 first Bytes as unsigned char (for info, but doesn't have any sense):"
AIMU.mode = "ALL"
print "Mode ALL:\t",[AIMU.value(i) for i in range(8)],
print struct.unpack('BBBBBBBB', AIMU.bin_data[:8])
```
    Mode TILT    :	[178, 138, 70] [178.20921534067105, 138.3895569113692, 69.63586519368219]
    Mode ACCEL   :	[944, 350, -103] [942, 350, -107]
    Mode COMPASS :	[319] 319
    Mode MAG     :	[1232, -121, 251] [1226, -122, 253]
    Mode GYRO    :	[1, 10, -8] [0.0, 12.4, -8.8]
    Show 8 first Bytes as unsigned char (for info, but doesn't have any sense):
    Mode ALL     :	[252, 171, 120, 176, 3, 94, 1, 149] (252, 171, 120, 176, 3, 94, 1, 149)

Maybe it would make more sense to remove all values when mode = ALL like it is described in the documentation. But that would be a new issue to see with @lechner :)

**Comment by ddemidov** [comment]
Thanks, now I see what is happening here.

**Comment by ddemidov** [comment]
Also, I believe the last comment was meant for @dlech :).

**Comment by cavenel** [comment]
My mistake...!

**Comment by dlech** [comment]
I love what you can do with Python. It makes programming so easy.

Anyway, you are on the right track here. For most modes of most sensors, the format of all the data values is the same, so the `bin_data_format` attribute will actually return a useful value. The IMU is a special case since it returns mixed data types and as a result it just returns unsigned bytes. And as you have figured out, it is up to the user program to interpret the bytes correctly.

Also in this case, as you saw, the `value<N>` attributes are the same as each byte. It is currently technically possible to have up to 32 values per mode, however, this is the only mode out of all of the supported sensors that has more than 8 values, so we currently just have 8 value attributes.

I agree that it makes sense to perhaps hide the value attributes or something like that to make this a bit more obvious. However, since this is a one-of-a-kind case, I'm not inclined to do anything about it.

**Comment by ddemidov** [comment]
Resolved by ev3dev/ev3dev-lang#84.
